addons/sourcemod/scripting/shavit-tas.sp

```diff
@@ -0,0 +1,816 @@
+	return Plugin_Continue;
+}
+
+public Action jointeam(int client, const char[] command, int args)
+{
+	if(g_bTAS[client]) {
+		gi_Status[client] = RUN;
+		gf_TASTime[client] = 0.0;
+		gi_IndexCounter[client] = 0;
+		FakeClientCommandEx(client, "sm_r");
+	}
+	return Plugin_Continue;
+}
+
+public Action Listener(int client, const char[] command, int args)
+{
+	if(!g_bTAS[client])
+	{
+		return Plugin_Continue;
+	}
+	if(StrEqual(command, "+rewind"))
+	{
+		gi_Status[client] = BACKWARD;
+		return Plugin_Handled;
+	}
+	else if(StrEqual(command, "+fastforward"))
+	{
+		gi_Status[client] = FORWARD;
+		return Plugin_Handled;
+	}
+	else if(StrEqual(command, "-rewind") || StrEqual(command, "-fastforward"))
+	{
+		gi_Status[client] = PAUSED;
+		return Plugin_Handled;
+	}
+	return Plugin_Continue;
+}
+
+public void OnClientPutInServer(int client)
+{
+	if(gh_Frames[client] != INVALID_HANDLE)
+		ClearArray(gh_Frames[client]);
+	else
+		gh_Frames[client] = CreateArray(11, 0);
+
+	gf_CounterSpeed[client] = 1.0;
+	gf_TASTime[client] = 0.0;
+	gf_TimeScale[client] = 1.0;
+	gi_Status[client] = RUN;
+	gb_TASMenu[client] = true;
+	AutoStrafeEnabled[client] = false;
+	g_Strafing[client] = false;
+}
+
+
+float GetClientVelo(int client)
+{
+	float vVel[3];
+	
+	vVel[0] = GetEntPropFloat(client, Prop_Send, "m_vecVelocity[0]");
+	vVel[1] = GetEntPropFloat(client, Prop_Send, "m_vecVelocity[1]");
+	
+	
+	return GetVectorLength(vVel);
+}
+
+public Action OnPlayerRunCmd(int client, int &buttons, int &impulse, float vel[3], float angles[3], int &weapon, int &subtype, int &cmdnum, int &tickcount, int &seed, int mouse[2])
+{
+	if (IsClientInGame(client) && !IsClientSourceTV(client) && !IsClientReplay(client) && IsClientConnected(client) && GetClientMenu(client) == MenuSource_None && g_bTAS[client] && IsPlayerAlive(client))
+	{
+		DrawPanel(client);
+	}
+
+	if(IsClientInGame(client))
+	{
+		if(!g_bTAS[client])
+		{
+			return Plugin_Continue;
+		}
+
+		if(Shavit_GetTimerStatus(client) != Timer_Running) {
+			return Plugin_Continue;
+		} else if(IsPlayerAlive(client) && !IsFakeClient(client)) {
+			if(gi_Status[client] == RUN) { // Record Frames
+				SetEntPropFloat(client, Prop_Send, "m_flLaggedMovementValue", gf_TimeScale[client]);
+				float fTimescale = GetEntPropFloat(client, Prop_Data, "m_flLaggedMovementValue");
+				gf_TimescaleTicksPassed[client] += fTimescale;
+				truevel[client] = GetClientVelo(client);
+				/*
+							AUTO STRAFER START
+														*/
+				if(buttons & IN_FORWARD && vel[0] <= 50.0)
+					vel[0] = 450.0;
+
+				float yaw_change = 0.0;
+				if(vel[0] > 50.0)
+					yaw_change = 30.0 * FloatAbs(30.0 / vel[0]);
+
+				if (AutoStrafeEnabled[client] == true && Shavit_GetTimerStatus(client) == Timer_Running && g_bTAS[client] && !(GetEntityFlags(client) & FL_ONGROUND) && (GetEntityMoveType(client) != MOVETYPE_NOCLIP) && !(buttons & IN_FORWARD) && !(buttons & IN_BACK) && !(buttons & IN_MOVELEFT) && !(buttons & IN_MOVERIGHT))
+				{
+					if(mouse[0] > 0)
```

Comment:
mouse isn't an accurate source for angle changes, requires players to run with joystick 0